test/src/unittests/rhythm/test_loopbpmconfidence.py

```diff
@@ -0,0 +1,157 @@
+# the terms of the GNU Affero General Public License as published by the Free
+# Software Foundation (FSF), either version 3 of the License, or (at your
+# option) any later version.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
+# details.
+#
+# You should have received a copy of the Affero GNU General Public License
+# version 3 along with this program. If not, see http://www.gnu.org/licenses/
+
+
+from essentia_test import *
+from essentia.standard import MonoLoader, LoopBpmConfidence
+
+class TestLoopBpmConfidence(TestCase):
+
+    def testInvalidParam(self):
+        self.assertConfigureFails(LoopBpmConfidence(), {'sampleRate': -1})
+
+    def testRegression(self):
+        audio = MonoLoader(filename=join(testdata.audio_dir, 'recorded', 'techno_loop.wav'))()        
+
+        """
+         Confidence figures are obtained from previous runs algorithm as of
+         code baseline on 10-Feb-2021
+         In keeping with regression tests principles, no future changes of algorithm
+         should break these tests.
+         Test for different BPMs starting with correct one(125),
+         and check the confidence levels obtained from previous runs.
+         The 3rd param in assertAlmostEqual() has been optimised to 0.02.
+        """        
+
+        bpmEstimate = 125
+        expectedConfidence = 1
+        confidence = LoopBpmConfidence()(audio, bpmEstimate)        
+        self.assertAlmostEqual(expectedConfidence, confidence, 0.02)
+
+        bpmEstimate = 120
+        expectedConfidence = 0.87
+        confidence = LoopBpmConfidence()(audio, bpmEstimate)        
+        self.assertAlmostEqual(expectedConfidence, confidence, 0.02)
+
+        bpmEstimate = 70
+        expectedConfidence = 0.68
+        confidence = LoopBpmConfidence()(audio, bpmEstimate)                
+        self.assertAlmostEqual(expectedConfidence, confidence, 0.02)
+
+        """
+         Apply same principle above to Sub-Sections of Audio.       
+         The following confidence measurements were taken on 11 Feb. 2021
+         for a given set of parameters on different audio subsections.
+         A segment of techno_loop gives a non-ideal estimation of BPM.
+         
+         Subsection [1000:6000], (Randomly chosen points, section 1)
+         BPM Estimate: 125: Confidence: 0.5275888442993164
+         BPM Estimate: 120: Confidence: 0.5464853048324585
+         BPM Estimate: 70: Confidence: 0.7354497313499451
+         Subsection [22000:38000], (Randomly chosen points, section 2)
+         BPM Estimate: 125: Confidence: 0.5117157697677612
+         BPM Estimate: 120: Confidence: 0.4512471556663513
+         BPM Estimate: 70: Confidence: 0.15343916416168213
+         The ASSERTs will check for these to two places of decimal
```

Comment:
90% of the loop results in the following estimates:
Subsection [0:len90]
         Estimate on the correct BPM:
         BPM Estimate: 125: Confidence: 0.19992440938949585
         Estimates on the incorrect BPMS:
         BPM Estimate: 120: Confidence: 0.4305669069290161
         BPM Estimate: 70: Confidence: 0.5011640191078186

Cutting1 % at beginning and 90% of the loop at the end results in the following estimates:         
Subsection [len01:len90],
         BPM Estimate: 125: Confidence: 0.9199735522270203
         BPM Estimate: 120: Confidence: 0.3631746172904968
         BPM Estimate: 70: Confidence: 0.7951852083206177

        I think the first test  (just 90%) is the best one to consider.
     
   Strangely , a high confidence results when we trim a bit at the beginning as well as the end.

Reply:
Hi, IIRC the algorithm considers 3 cases in which a loop is not cut correctly but we still want to provide high confidence if the tempo estimation is good. This is when i) there is *non-musical** silence before the loop starts, ii) thee is non-musical silence after the loop ends, iii) there is non-musical silence at the start and AND the end of the loop. Because the algorithm matches the duration of the sound with the estimate of the BPM, non-musical silence at these points would invalidate the confidence value. What the algorithm does is to estimate the effective start and end points of the loop and also use these "alternative" durations in the comparison with the estimate of the BPM. I think the highest confidence is returned.

The algorithm does not account for cases in which the loop is cut in the middle (although if the loop is cut at a multiple of the estimated BPM beat duration, then the confidence should be high as well). I hope this clarifies.


* by non-musical I mean that silence is not part of the loop itself.

Reply:
@ffont Thanks! The documentation for LoopBpmConfidence could be improved to mention that we consider start/end silences. For now it only says "The confidence returned is based on comparing the duration of the signal with multiples of the BPM estimate." This can be misleading.

Reply:
@cvf-bcn-gituser @ffont We should be fine with using a single 90% segment for the regression test in LoopBpmEstimator, but should capture the 3 cases that can happen in test for LoopBpmConfidence. We can simulate them by adding artificial silence in the beginning or the end of audio loop. As I understand, given a BPM value, a certain amount of added silence should not change the output confidence unless it makes the new total audio length a better fit for the BPM value.

Does this imply that given a BPM value, adding artificial silence in the beginning or the end of the audio loop should not affect the resulting confidence value in LoopBpmConfidence?

Reply:
> @ffont Thanks! The documentation for LoopBpmConfidence could be improved to mention that we consider start/end silences. For now it only says "The confidence returned is based on comparing the duration of the signal with multiples of the BPM estimate." This can be misleading.

Ok, we can add the sentence: "The confidence returned is based on comparing the duration of the signal with multiples of the BPM estimate. The algorithm considers edge cases in which the signal includes non-musical silence at the start or end of the file (or both) and returns the confidence for the best fit." 
However we should refer to the paper for details, it is hard to give more details in the docstring. This is the link to the paper in case it is not there: https://repositori.upf.edu/handle/10230/33113

Reply:
> We can simulate them by adding artificial silence in the beginning or the end of audio loop. As I understand, given a BPM value, a certain amount of added silence should not change the output confidence unless it makes the new total audio length a better fit for the BPM value.

Yes, I think this is how it works. The test would be that adding artificial silence at the start of the signal (or at the end or at both places) gives the same result as not adding any artificial silence.

Reply:
This paper is mentioned in the DOC. The new phrase looks very good, the only thing I would change is "file2 -> "audio signal"